client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-ai-assembler.ts

```diff
@@ -0,0 +1,74 @@
+import { useState } from 'react';
+import { useSearchParams } from 'react-router-dom';
+import wpcomRequest from 'wpcom-proxy-request';
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+export default function useAIAssembler(): [ Function, Function, string, boolean ] {
+	const [ searchParams, setSearchParams ] = useSearchParams();
+	const [ prompt, setPrompt ] = useState( searchParams.get( 'ai_description' ) || '' );
+	const [ loading, setLoading ] = useState( false );
+
+	function callAIAssembler() {
+		setSearchParams( ( currentSearchParams: any ) => {
+			currentSearchParams.set( 'ai_description', prompt );
+			return currentSearchParams;
+		} );
+		setLoading( true );
+		return wpcomRequest( {
+			path: '/pattern-assembler/ai/latest/generate',
+			method: 'POST',
+			apiNamespace: 'wpcom/v2',
+			body: {
+				description: prompt,
+			},
+		} )
+			.catch( ( err ) => {
+				setLoading( false );
+				console.log( 'Patterns AI error', err ); /* eslint-disable-line no-console */
+				return Promise.reject( err );
+			} )
+			.then( ( response: any ) => {
+				setLoading( false );
+				console.log( 'Patterns AI response', response ); /* eslint-disable-line no-console */
+				// This actually passes the patterns to the pattern assembler.
+				setSearchParams(
+					( currentSearchParams: any ) => {
+						currentSearchParams.set( 'header_pattern_id', response.header_pattern );
+						currentSearchParams.set( 'footer_pattern_id', response.footer_pattern );
+						currentSearchParams.set( 'pattern_ids', response.pages[ 0 ].patterns.join( ',' ) );
+						// These 2 params were introduced in the V5 of the AI endpoint.
+						// TODO: Remove the checks once we settle on the endpoint.
+						if ( response?.site?.site_title ) {
+							currentSearchParams.set( 'site_title', response.site.site_title );
+						}
+						if ( response?.site?.site_tagline ) {
+							currentSearchParams.set( 'site_tagline', response.site.site_tagline );
+						}
+
+						// This was introduced in the V5 of the AI endpoint.
+						const pageSlugs = response.pages.map( ( page: any ) => page?.slug ).filter( Boolean );
+						if ( pageSlugs.length ) {
+							currentSearchParams.set( 'page_slugs', pageSlugs.join( ',' ) );
+						}
+
+						if ( response.style ) {
+							currentSearchParams.set( 'color_variation_title', response.style );
+						}
+
+						if ( response.font ) {
+							currentSearchParams.set( 'font_variation_title', response.font );
+						}
+
+						// So that we close the drawer with patterns when moving to the assembler:
+						currentSearchParams.set( 'screen', 'main' );
+						currentSearchParams.delete( 'screen_parameter' );
+
+						return currentSearchParams;
+					},
+					{ replace: true }
+				);
+				return Promise.resolve( response );
+			} );
+	}
+	return [ callAIAssembler, setPrompt, prompt, loading ];
```

Comment:
[nit] Could we return an object instead of an array, so that it's order agnostic?

Reply:
Yeah, I was thinking about this but was unsure if this is the preferred way, as doing it as an object requires you to keep the names, which can conflict with other variables in the file you are using this effect in.

I can try and change this next time I'm touching these files